frontend/controller/serviceworkers/sw-primary.js

```diff
@@ -83,36 +223,73 @@ self.addEventListener('message', function (event) {
 // Handle clicks on notifications issued via registration.showNotification().
 self.addEventListener('notificationclick', event => {
   console.debug('[sw] Notification clicked:', event.notification)
-})
+  event.notification.close()
 
-self.addEventListener('push', function (event) {
-  // PushEvent reference: https://developer.mozilla.org/en-US/docs/Web/API/PushEvent
+  event.waitUntil(
+    self.clients.matchAll({ type: 'window' }).then((clientList) => {
+      clientList.sort((a, b) => {
+        if (a.focused !== b.focused) {
+          return a.focused ? -1 : 1
+        }
+        if (a.visibilityState !== b.visibilityState) {
+        // order is visible, prerender, hidden
+          if (a.visibilityState === 'visible') return -1
+          if (b.visibilityState === 'visible') return 1
+          if (a.visibilityState === 'hidden') return 1
+          if (b.visibilityState === 'hidden') return -1
+        }
 
-  if (!(self.Notification && self.Notification.permission === 'granted')) {
-    console.debug("[sw] received a push notification but aren't displaying it due to the permission not granted")
-    return
-  }
-
-  const data = event.data.json()
-  console.debug('[sw] push received: ', data)
+        return 0
+      })
+      if (!clientList.length) {
+        return self.clients.openWindow(`/app${event.notification.data.path ?? '/'}`)
+      }
+      const client = clientList[0]
+      if (event.notification.data?.path) {
+        client.postMessage({
+          type: 'navigate',
+          path: event.notification.data.path
+        })
+      }
+      if (!client.focused) return client.focus()
+    }))
+})
 
-  self.registration.showNotification(
-    data.title,
-    {
-      body: data.body || '',
-      icon: '/assets/images/pwa-icons/group-income-icon-transparent.svg'
+sbp('okTurtles.events/on', KV_EVENT, ({ contractID, key, data }) => {
+  const rootState = sbp('chelonia/rootState')
+  switch (key) {
+    case KV_KEYS.LAST_LOGGED_IN: {
+      if (!rootState.lastLoggedIn) rootState.lastLoggedIn = Object.create(null)
+      rootState.lastLoggedIn[contractID] = data
+      break
+    }
+    case KV_KEYS.UNREAD_MESSAGES: {
+      if (!rootState.chatroom) rootState.chatroom = Object.create(null)
+      rootState.chatroom.unreadMessages = data
+      break
+    }
+    case KV_KEYS.PREFERENCES: {
+      rootState.preferences = data
+      break
     }
-  )
+    case KV_KEYS.NOTIFICATIONS: {
+      if (!rootState.notifications) rootState.notifications = Object.create(null)
+      rootState.notifications.status = data
+      break
+    }
```

Comment:
Can we add a `default` here with a `console.error` or `console.warn` to indicate an unknown event was received (along with the event name and data)?

Reply:
I'm not sure that'd be helpful. The logs already include these events (via `/emit`) and not all keys necessarily need to be processed here (although they currently are). Other contracts can also define arbitrary KV keys that may not need to be processed. If we do add a log line, I think it shouldn't be at the `warn` or `error` level; the `info` or even `debug` levels seem more appropriate (since the main use of logging this would be a dev mistyping a key, for which ideally a constant should be used anyhow).